### DISCUSSION TITLE
Global concurrency per queue

### DISCUSSION THREAD

**Discussion by marklynk** [taskforcesh/bullmq#830] [Q&A]
Hello! Is it possible to set a global concurrency limit on workers? For example, let's say I create two workers for redundancy sake, but I don't care how many jobs each one processes at a time individually (the real work is kicked off to an external resource).

Is it possible to set a global concurrency limit such that if I have 2 workers running with a global concurrency of 10, worker **A** might happen to execute 3 jobs and worker **B** might execute 7. But, if worker **A** goes down, then for sure worker **B** will be able to execute 10.

**Comment by manast** [comment]
That's how rate limit works already. You define the rate limit when you add the job and then this is preserved by the workers.

**Reply by nemphys** [reply]
I still don't understand how the requested feature (global queue concurrency) can be achieved through the rate limit (worker "limiter") setting.

Let's say I have 12 workers, each with a concurrency setting of 1, but I only want a maximum of 4 jobs executed simultaneously at any given time (by any number of workers). I tried setting { limiter: { max: 4, duration: 1000 } } on all workers, but when I add 20 jobs to the queue, I can see 12 jobs being executed at the same time.

**Reply by manast** [reply]
The rate limit limits the rate at which jobs are being picked from the queue by the workers, not the concurrency of the workers. So if you have slow jobs it is very possible that you can have 12 jobs being processed simultaneous while having a limit of 4 jobs per second.

**Reply by nemphys** [reply]
Yes, that is exactly my point, since a job can take anywhere between 1 - 30 seconds in my case.
What I (and @edgmark) need is a concurrency limit *per queue*, which appears to be impossible at the moment.

**Reply by manast** [reply]
But the concurrency depends on the amount of workers and concurrency factors that you have configured yourself... If you want to have more concurrency on the workers and more workers and still limit the concurrency then you can use the max concurrency feature for groups in the Pro version: https://docs.bullmq.io/bullmq-pro/groups/concurrency

**Reply by nemphys** [reply]
Right, I suppose this is a pro-only feature then, no problem with that. I just wanted to make sure that I had not misunderstood something.
